docs/RFCS/version_migration.md

```diff
@@ -0,0 +1,155 @@
+the prior version, or else the operator may confirm the new version as
+the minimum required version, allowing new, backwards-incompatible
+features to run.
+
+# Motivation
+
+Introducing backwards-incompatible features is currently risky, as
+usage could prevent orderly downgrades, or worse, could result in
+inconsistencies if used during a rolling upgrade.
+
+One example is the use of different storage for Raft's write ahead log
+(#16809). Having nodes begin using this new functionality means they
+cannot be downgraded without potential loss of data. To address this,
+CockroachDB migration must provide a transition period where all nodes
+upgrade to the version which supports the new Raft WAL storage, but do
+not yet use it. At this point, the cluster can still be downgraded to
+the prior version, allowing the operator to safely test the new
+build. When satisfied with the new version's stability, the operator
+confirms that the new version should become the minimum required, at
+which point the nodes begin using the Raft WAL storage, and the window
+for downgrading is closed.
+
+Another example is the `Revert` command. Once executed, it requires
+that all nodes performing reads understand how to utilize new
+information to ignore reverted data during reads. If one replica out
+of three were not upgraded to understand the effects of a `Revert`,
+then that node could incorrectly read reverted data as if it were
+still live.
+
+Backwards-incompatible changes to *existing* functionality are even
+more problematic because you'd expect them to be in active use at the
+time of upgrade. In this case, there would be no way to do a rolling
+upgrade without having some nodes with the old version and some nodes
+with the new, backwards-incompatible version.
+
+We expect most point releases to contain features requiring the
+migration support described in this RFC.
+
+# Detailed design
+
+Every node knows the version it's running, as the version is baked
+into the build. A node's version will be supplied with every RPC
+(e.g. Raft, Node, DistSQL, and Gossip). In addition, there will be a
+cluster-wide minimum required version, set by the operator. Each node
+also sends the minimum required version its operating under with each
+RPC.
+
+- RPCs from nodes with versions older than the minimum required
+  version will fail with an error that indicates the sender should
+  panic.
+- If an RPC is received which specifies a minimum required version
+  which is newer than the receiver's minimum required version, an
+  error is returned that indicates the sender should backoff and retry
+  (while the receiver waits for new gossip).
+- If an RPC is received which specifies a minimum required version
+  which is newer than the receiver's version, the receiver returns
+  an error and panics.
+
+The minimum required version will be stored and gossiped as a cluster
+setting. The value of the minimum required version is stored in a
```

Comment:
It's not clear to me why we need a protobuf as opposed to (say) a string with semver format.